src/controllers/krate/delete.rs

```diff
@@ -0,0 +1,420 @@
+            .first::<i32>(&mut conn)
+            .await
+            .optional()?
+            .is_some();
+
+        if has_rev_dep {
+            let msg = "only crates without reverse dependencies can be deleted after 72 hours";
+            return Err(custom(StatusCode::UNPROCESSABLE_ENTITY, msg));
+        }
+    }
+
+    conn.transaction(|conn| {
+        async move {
+            // Delete the crate
+            diesel::delete(crates::table.find(krate.id))
+                .execute(conn)
+                .await?;
+
+            // Enqueue index sync background jobs
+            jobs::SyncToGitIndex::new(&krate.name)
+                .async_enqueue(conn)
+                .await?;
+
+            jobs::SyncToSparseIndex::new(&krate.name)
+                .async_enqueue(conn)
+                .await?;
+
+            // Enqueue deletion of corresponding files from S3
+            jobs::DeleteCrateFromStorage::new(name)
+                .async_enqueue(conn)
+                .await?;
+
+            Ok::<_, BoxedAppError>(())
+        }
+        .scope_boxed()
+    })
+    .await?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::models::OwnerKind;
+    use crate::tests::builders::{DependencyBuilder, PublishBuilder};
+    use crate::tests::util::{RequestHelper, Response, TestApp};
+    use crates_io_database::schema::crate_owners;
+    use diesel_async::AsyncPgConnection;
+    use http::StatusCode;
+    use insta::assert_snapshot;
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn test_happy_path_new_crate() -> anyhow::Result<()> {
+        let (app, anon, user) = TestApp::full().with_user();
+        let mut conn = app.async_db_conn().await;
+        let upstream = app.upstream_index();
+
+        publish_crate(&user, "foo").await;
+        let crate_id = adjust_creation_date(&mut conn, "foo", 71).await?;
+
+        // Update downloads count so that it wouldn't be deletable if it wasn't new
+        adjust_downloads(&mut conn, crate_id, DOWNLOADS_PER_MONTH_LIMIT * 100).await?;
+
+        assert_crate_exists(&anon, "foo", true).await;
+        assert!(upstream.crate_exists("foo")?);
+        assert_snapshot!(app.stored_files().await.join("\n"), @r"
+        crates/foo/foo-1.0.0.crate
+        index/3/f/foo
+        rss/crates.xml
+        rss/crates/foo.xml
+        rss/updates.xml
+        ");
+
+        let response = delete_crate(&user, "foo").await;
+        assert_eq!(response.status(), StatusCode::OK);
+        assert!(response.body().is_empty());
+
+        // Assert that the crate no longer exists
+        assert_crate_exists(&anon, "foo", false).await;
+        assert!(!upstream.crate_exists("foo")?);
+        assert_snapshot!(app.stored_files().await.join("\n"), @r"
+        rss/crates.xml
+        rss/updates.xml
+        ");
+
+        Ok(())
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn test_happy_path_old_crate() -> anyhow::Result<()> {
```

Comment:
Since there is a per-month condition, I suggest adding a test case that is older than a month.